components/calendar/calendar.js

```diff
@@ -0,0 +1,314 @@
+import React, { Component, PropTypes } from 'react';
+import { getClassNamesWithMods, getDataAttributes, normalizeDate } from '../_helpers';
+import DaysPanel from './panels/days';
+import calendarConstants from './constants/calendar';
+
+const {
+  CALENDAR_MOVE_TO_NEXT,
+  CALENDAR_MOVE_TO_PREVIOUS,
+  CALENDAR_SELECTION_TYPE_RANGE,
+} = calendarConstants;
+
+
+/**
+ * Processes the given props and the existing state and returns
+ * a new state.
+ *
+ * @function processProps
+ * @param {Object} props Props to base the new state on.
+ * @param {Object} state (Existing) state to be based on for the existing values.
+ * @return {Object} New state to be set/used.
+ * @static
+ */
+function processProps(props) {
+  const { initialDates, maxDate, minDate, selectionType } = props;
+  const maxLimit = maxDate ? normalizeDate(new Date(maxDate), 23, 59, 59, 999) : null;
+  const renderDate = normalizeDate(((initialDates && initialDates.length && initialDates[0])
```

Comment:
I agree with you about what you mentioned about side-effects. The only problem we have here is that it is still referencing the parent Component's props.

I would always avoid props mutation of any kind in React. If I really need to transform the value in some way, and expect the Component to re-render on further changes, I would use `.setState()`, and scope it under the current Component only. Without affecting parent(s).

Similar issue was found in another PR few weeks ago. I think @asci wanted to created a Story out of it so all affected components are updated to avoid this issue.